PR: https://github.com/odoo/odoo/pull/273948

From: c5037bbe07890ec389b7a1db057c2d2edbb075f0
From: Nisarg (nipl)
Date: 2026-09-09 06:21:38

Breaking data model changes scores: del:1 + add:0, change matches:
-    translated_product_name = fields.Text(compute='_compute_translated_product_name')

Total Changes: 101

[IMP] sale,account,purchase,*: remove product name from description

Sale order lines, purchase order lines, and account move lines currently
store the product's display name together with its description in the
`name` field. This behavior was originally introduced so the product
field could remain hidden while still displaying both the product and
its description in a single field.

Maintaining this behavior requires a significant amount of JS logic and
has led to several inconsistencies across the different models.

Now that the framework supports stacking multiple fields in the same
column, this workaround is no longer necessary.

Keeping the product name in the description also has several drawbacks:
* Product names are unnecessarily exported as EDI descriptions.
* The description field always contains a value, making it impossible to
  tell whether a custom description was actually provided.
* Models such as sale order templates do not follow the same convention,
  resulting to complex logic to make it work.
* The JS widget needs complex logic to strip the product name while
  taking translations into account.
* The product/description logic has become tightly coupled with the
  sections widget, making it difficult to maintain.

This commit stores only the product description in the `name` field and
introduces a new non-stored computed `label` field on sale order lines,
purchase order lines, and account move lines. The `label` field combines
the product display name and description and is used wherever both
pieces of information need to be displayed, such as reports, the portal,
and list view seachable text field introduced in d069ce5.

For sale, the extra logic required to preserve the product name across
sale orders and sale order templates is removed, allowing the `name`
field to be used consistently as the product description.

For purchase, the seller-specific naming logic is removed from the
compute methods and delegated to the product display name retrieval.

For accounting, EDI exports now contain only the actual line description
instead of the product name followed by the description.

This commit is the first step towards adapting the new framework
`<column>` support for the product and description field, allowing the
remaining JS logic to be significantly simplified in follow-up commits.

task-6241473

closes odoo/odoo#273948

Related: odoo/enterprise#122792
Related: odoo/upgrade#10919
Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/purchase/models/account_invoice.py
+++ b/addons/purchase/models/account_invoice.py
@@ -435,7 +435,10 @@ class AccountMove(models.Model):
                         # account.move.line model with the extra matching fields and extending the EDI extraction
                         # logic to fill these new fields.
                         purchase_line_candidates[purchase_line] = difflib.SequenceMatcher(
-                            None, invoice_line.name, purchase_line.name).ratio()
+                            None,
+                            invoice_line.label,
+                            purchase_line.label,
+                        ).ratio()
 
                 if len(purchase_line_candidates) > 0:
                     # We take the best match based on the name.

--- a/addons/purchase/models/purchase_order_line.py
+++ b/addons/purchase/models/purchase_order_line.py
@@ -20,8 +20,8 @@ class PurchaseOrderLine(models.Model):
     _order = 'order_id, sequence, id'
 
     name = fields.Text(
-        string='Description', required=True, compute='_compute_price_unit_and_date_planned_and_name', store=True, readonly=False)
-    translated_product_name = fields.Text(compute='_compute_translated_product_name')
+        string='Description', compute='_compute_price_unit_and_date_planned_and_name', store=True, readonly=False)
+    label = fields.Text(string="Label", compute="_compute_label", inverse="_inverse_label")
     sequence = fields.Integer(string='Sequence', default=10)
     product_qty = fields.Float(string='Quantity', digits='Product Unit', required=True)
     product_uom_qty = fields.Float(string='Total Quantity', compute='_compute_product_uom_qty', store=True)
@@ -482,34 +482,15 @@ class PurchaseOrderLine(models.Model):
         for line in self:
             if not line.product_id or line.invoice_lines or not line.company_id or self.env.context.get('skip_uom_conversion') or (line.technical_price_unit != line.price_unit):
                 continue
-            params = line._get_select_sellers_params()
+
             seller_info = line._get_seller_info()
+
             if seller_info or not line.date_planned:
                 line.date_planned = line._get_date_planned(seller_info).strftime(DEFAULT_SERVER_DATETIME_FORMAT)
 
-            # record product names to avoid resetting custom descriptions
-            default_names = []
-            display_names = []
-            vendors = line.product_id._prepare_sellers(params=params)
-            product_ctx = {'seller_id': None, 'partner_id': None, 'lang': get_lang(line.env, line.partner_id.lang).code}
-            line_without_seller = line.product_id.with_context(product_ctx)
-            default_names.append(line._get_product_purchase_description(line_without_seller))
-            for vendor in vendors:
-                product_ctx = {'seller_id': vendor.id, 'lang': get_lang(line.env, line.partner_id.lang).code}
-                default_names.append(line._get_product_purchase_description(line.product_id.with_context(product_ctx)))
-                display_names.append(line.product_id.with_context(product_ctx).display_name)
-            if not line.name or line.name in default_names:
-                product_ctx = {'seller_id': line.selected_seller_id.id, 'lang': get_lang(line.env, line.partner_id.lang).code}
+            if not line.name:
+                product_ctx = {'lang': get_lang(line.env, line.partner_id.lang).code}
                 line.name = line._get_product_purchase_description(line.product_id.with_context(product_ctx))
-            else:
-                # Checks that the product vendor and vendor name are correct
-                for vendor, display_name in zip(vendors, display_names):
-                    if line.name.startswith(display_name):
-                        if not line.selected_seller_id:
-                            line.name = line_without_seller.display_name + line.name[len(display_name):]
-                        elif vendor.id != line.selected_seller_id.id:
-                            line.name = display_names[vendors.ids.index(line.selected_seller_id.id)] + line.name[len(display_name):]
-                        break
 
             # If no seller, use the standard price. It needs a proper currency conversion.
             if not seller_info:
@@ -560,12 +541,45 @@ class PurchaseOrderLine(models.Model):
             'technical_price_unit': price_unit,
         })
 
-    @api.depends('product_id')
-    def _compute_translated_product_name(self):
+    @api.depends("product_id", "selected_seller_id", "partner_id.lang", "name")
+    def _compute_label(self):
+        for line in self:
+            if not line.product_id:
+                line.label = line.name
+                continue
+
+            product_display_name = line._get_product_display_name()
+            if not line.name:
+                line.label = product_display_name
+            elif line.name.splitlines()[0] == product_display_name:
+                # If description already holds the product name, use it as label
+                line.label = line.name
+            else:
+                line.label = product_display_name + "\n" + line.name
+
+    def _inverse_label(self):
         for line in self:
-            line.translated_product_name = line.product_id.with_context(
-                lang=line.partner_id.lang,
-            ).display_name
+            display_name = line._get_product_display_name()
+
+            if display_name and line.label:
+                line.name = (
+                    line.label
+                    .removeprefix(display_name)
+                    .removeprefix("\n")
+                )
+            else:
+                line.name = line.label
+
+    def _get_product_display_name(self):
+        self.ensure_one()
+        if not self.product_id:
+            return ""
+
+        product = self.product_id.with_context(
+            seller_id=self.selected_seller_id.id,
+            lang=get_lang(self.env, self.partner_id.lang).code,
+        )
+        return product.display_name
 
     @api.depends('uom_id', 'product_qty', 'product_id.uom_id')
     def _compute_product_uom_qty(self):
@@ -684,13 +698,16 @@ class PurchaseOrderLine(models.Model):
 
     def _get_product_purchase_description(self, product_lang):
         self.ensure_one()
-        name = product_lang.display_name
-        if product_lang.description_purchase:
-            name += '\n' + product_lang.description_purchase
-        product_lang_no_variant_attribute_value_ids = self.with_context(product_lang.env.context).product_no_variant_attribute_value_ids
+        name = product_lang.description_purchase or ""
+        product_lang_no_variant_attribute_value_ids = self.with_context(
+            product_lang.env.context
+        ).product_no_variant_attribute_value_ids
         for no_variant_attribute_value in product_lang_no_variant_attribute_value_ids:
-            name += "\n" + no_variant_attribute_value.attribute_id.name + ': ' + no_variant_attribute_value.name
-
+            if name:
+                name += "\n"
+            name += (
+                f"{no_variant_attribute_value.attribute_id.name}: {no_variant_attribute_value.name}"
+            )
         return name
 
     def _prepare_account_move_line(self, move=False):
@@ -700,7 +717,7 @@ class PurchaseOrderLine(models.Model):
 
         res = {
             'display_type': self.display_type or 'product',
-            'name': self.env['account.move.line']._get_journal_items_full_name(self.name, self.product_id.display_name),
+            'name': self.name,
             'product_id': self.product_id.id,
             'product_uom_id': self.uom_id.id,
             'quantity': -self.qty_to_invoice if move and move.move_type == 'in_refund' else self.qty_to_invoice,
@@ -764,15 +781,12 @@ class PurchaseOrderLine(models.Model):
             lang=partner_id.lang,
             partner_id=partner_id.id,
         )
-        name = product_lang.with_context(seller_id=seller_info['supplierinfo'].id if seller_info else False).display_name
-        if product_lang.description_purchase:
-            name += '\n' + product_lang.description_purchase
 
         date_planned = self.order_id.date_planned or self._get_date_planned(seller_info, po=po)
         discount = seller_info.get('discount') or 0.0
 
         return {
-            'name': name,
+            'name': product_lang.description_purchase or '',
             'product_qty': product_qty if product_uom else uom_po_qty,
             'product_id': product_id.id,
             'uom_id': product_uom.id or seller_uom.id,
